상품 브랜드 캐시 적용

Hosick opened the pull request:
- Redis를 이용한 글로벌 캐싱 설정
- 브랜드 전체 조회시 캐시 등록
- 브랜드 생성, 수정, 삭제 시 브랜드 List 캐시 삭제
- 상품 조회시 캐시 등록
- 상품 수정, 삭제 시 상품 캐시 삭제

Hosick committed:
feat: 캐시 적용 선행작업

-상품 DTO 내 출시일 필드에 LocalDateSerializer 적용
-상품 브랜드 DTO에 기본 생성자 추가
-브랜드, 상품 컨트롤러에 관리자 체크 미리 주석처리

issue number: #30

Hosick committed:
feat: 브랜드 및 상품 캐시 적용

-Redis를 이용한 글로벌 캐싱 설정
-브랜드, 상품 수정 메소드 트랜잭션 적용
-브랜드 전체 조회시 캐시 등록
-브랜드 생성, 수정, 삭제 시 브랜드 List 캐시 삭제
-상품 조회시 캐시 등록
-상품 수정, 삭제 시 상품 캐시 삭제

issue number: #30

Hosick commented:
브랜드 전체 조회와 상품 정보 조회는 유저가 접근할 수 있는 기능입니다.
두 기능 모두 상대적으로 변경이 잦지 않으며, 트래픽도 어느정도 있을 것이라고 예상하여 캐시 적용하였습니다.

 다만 곧 구현할 거래 도메인(Trade)이 상품과 양방향 연관관계를 맺게 되면, 상품 조회 페이지에 상품 뿐 아니라 상품의 거래 내역이 조회되게 됩니다. 이 때 거래 List는 상대적으로 변경이 잦을 텐데 그렇다면 상품과 거래를 묶어서 캐싱하는 방법과 따로 캐싱하는 방법 중 어느 방법이 나을지 고민됩니다.

제 생각에는 상품과 거래를 따로 조회하여 캐싱한다면 Evict 발생 시 거래(또는 상품)만 다시 조회해도 되기 때문에 더 좋은 방법같은데, 조인 쿼리로 조회하는 것이 아니기 때문에 쿼리가 2회 발생할 수 있다는 점이 조금 걸립니다.

Hosick commented:
하나 더 궁금한 점이 있는데 캐싱할 대상을 선정할 때 변경이 잦지 않고, 트래픽이 많은 데이터를 선정하려고 하는데,
거래 리스트의 경우 인기가 많은 제품의 거래는 변경이 잦고, 인기가 없는 제품의 거래는 변경이 드물텐데 
이런 경우에는 어떻게 선정하면 좋을까요?

f-lab-dev commented:
질문이 좀 정리되면 좋을 것 같은데

1. 양방향 연관관계를 꼭 맺어야하는지 고민해주시고요~
2. 거래를 캐싱할 필요가 있을지, 그리고 상품과 거래 데이터의 일관성에 대해서 고민을 해보시면 좋을 것 같아요
3. 무조건 `변경이 잦지 않고, 트래픽이 많다`가 캐싱의 조건은 아닙니다. 데이터마다, 동작마다 특성이 다르기때문에 여기도 고민이 많이 필요할거고요.
4. `cache eviction algorithms` 에 대해 알아보시면 좋을 것 같아요

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/config/CacheConfig.java:
```diff
@@ -0,0 +1,59 @@
+/**
+ * @EnableCaching :@Cacheable, @CacheEvict 등의 캐시 어노테이션 활성화를 위한 어노테이션이다. AOP 를 이용한 캐싱 기능을 추상화 시켜주므로
+ * 별도의 캐시 관련 로직없이 간단하게 적용해 줄 수 있다. 프록시는 Spring AOP 를 이용하여 동작하며 mode 설정을 통해 aspectj 를 이용한 방법으로 설정해 줄
+ * 수도 있다.
```
핵심만 가져와보면 `캐시 어노테이션 활성화`인데 스프링 내부에서 어떻게 이런 캐싱로직이 추상화되어있고, 어떤 로직을 타게되는지 알아보시면 좋을 것 같아요~

Hosick replied:
`@EnableCaching` 어노테이션을 적용하면 모든 스프링 빈에서 public 메소드에 캐싱 어노테이션이 있는지 검사하는 빈 후처리기를 트리거 합니다. 그 후 동적 프록시를 이용하여 어노테이션이 붙은 메소드의 호출을 가로채고 캐싱 동작을 수행합니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/config/CacheConfig.java:
```diff
@@ -0,0 +1,59 @@
+            .serializeKeysWith(RedisSerializationContext.SerializationPair
+                .fromSerializer(new StringRedisSerializer()))
+            .serializeValuesWith(RedisSerializationContext.SerializationPair
+                .fromSerializer(new GenericJackson2JsonRedisSerializer()));
```
1. 이 `Serializer`는 어떤 것일까요?
2. `GenericJackson2JsonRedisSerializer`는 어떤 클래스일까요?

Hosick replied:
Redis의 저장 형식이 byte array이므로 key와 value를 직렬화 해주어야합니다.
GenericJackson2JsonRedisSerializer는 객체를 Json으로 변환하여 직렬화 해주는 역할을 하는 것으로 알고있습니다!

f-lab-dev replied:
객체를 JSON으로 직렬화해주는거 외에 특징이 더 있습니다~

Hosick replied:
`GenericJackson2JsonRedisSerializer`는 직렬화 시 타입 정보 또한 함께 저장하기 때문에 클래스를 따로 지정해주지 않아도 역직렬화가 가능합니다. 따라서 여러 클래스를 범용적으로 직렬화 해줄 필요가 있는 경우 사용하기 좋습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/BrandApiController.java:
```diff
@@ -44,12 +46,14 @@
         return OK;
     }
 
+    //@LoginCheck(authority = admin)
```
주석으로 두기보다는 빼는게 낫지 않을까요? 꼭 필요한거라면 TODO로 주석으로 처리한 이유를 적어두시면 좋을 것 같아요

Hosick replied:
주석을 제거하고 추후 관리자 기능이 merge되었을 때 다시 추가하도록 하겠습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/controller/dto/ProductDto.java:
```diff
@@ -85,14 +89,18 @@ public Product toEntity() {
         private String modelNumber;
         private String color;
+        @JsonDeserialize(using = LocalDateDeserializer.class)
+        @JsonSerialize(using = LocalDateSerializer.class)
```
이렇게 데이터마다 붙이는게 아니라 공통적으로 처리할 수 있는 방법이 없을까요?

Hosick replied:
Serializer에 적절한 ObjectMapper를 정의하고 적용하여 공통처리로 변경하였습니다!

f-lab-dev commented on src/main/java/com/flab/shoeauction/service/BrandService.java:
```diff
@@ -38,21 +42,26 @@ public BrandInfo getBrandInfo(Long id) {
             .toBrandInfo();
     }
 
+    @Cacheable(value = "brands")
```
이 캐시의 만료시간은 어떻게 되나요?

Hosick replied:
`RedisCacheConfiguration.defaultConfig`는 기본적으로 만료시간이 정해져 있지 않은 것으로 알고 있습니다.
하지만 그렇다면 인기가 없는 상품같은 경우에 조회되지 않는데도 메모리만 차지할 수도 있다고 생각됩니다.
브랜드 목록과 상품 정보의 기본 TTL을 24시간으로 설정하도록 하겠습니다!

f-lab-dev replied:
캐시마다 따로 만료시간을 정해둘 수 있지 않을까요?

Hosick replied:
ttl이 다른 RedisCacheManager를 만들고 `@Casheable`마다 manager를 따로 적용해주면 될 것 같습니다.
데이터 별로 적절한 만료시간을 정하는 것이 헷갈려서 함께 처리했었는데 따로 적용하도록 수정하겠습니다!

f-lab-dev replied:
캐시매니저를 너무 많이 생성하게될 것 같은데 다른 방법도 찾아보시면 좋을 것 같아요~

Hosick replied:
앗 처음에 캐시 매니저를 여러개 Bean으로 등록하는 방법을 생각했다가  매니저가 계속 늘어날 것 같다는 생각에
외부 주입으로 캐시 별 Timeout을 받고 withInitialCacheConfigurations에 여러 Configuration들이 담긴 HashMap을 넣어주는 방법으로 변경하였습니다.

kihyuk-jeong commented on src/main/java/com/flab/shoeauction/service/BrandService.java:
```diff
@@ -38,21 +42,26 @@ public BrandInfo getBrandInfo(Long id) {
         brandRepository.deleteById(id);
     }
 
+    @CacheEvict(value = "brands", allEntries = true)
```
혹시 brand 에는 key 값을 주는대신  `allEntries = true` 를 사용하시고 product에는 `key`값을 부여하신 이유가 있으실까요?

Hosick replied:
브랜드 전체 조회는 key가 따로 없고 하나로 캐싱되기 때문에 allEntries를 true로 하여 전부 evict하여도 상관없지만, 
상품 정보 조회는 id를 key로 하여 따로 캐싱되기 때문에 정해진 key만 evict 해주어야 합니다!

Hosick committed:
feat: 캐시 별 만료기간 설정

- properites에서 캐시 만료기간 설정들을 Map으로 바인딩 받는
CacheProperties 구현
- RedisCache 기본 설정 함수 분리
- RedisCacheManager에 InitialCacheConfigurations 설정

- 관리자 기능 주석 제거

issue number: #30

Hosick committed:
feat: 리뷰반영

- LocalDate 직렬화를 위한 ObjectMapper 등록
- 주석 수정

issue number: #30

Hosick committed:
feat: 상품 생성 시 브랜드 검사 추가

-브랜드 검사 로직 추가

issue number: #30

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/config/CacheProperties.java:
```diff
@@ -0,0 +1,28 @@
+@Getter
+@Setter
+@Component
+@ConfigurationProperties(prefix = "redis-cache-constants")
```
이런 properties에는 보통 `constants`라는 단어를 쓰지 않습니다~

Hosick replied:
Constants는 어플리케이션 내의 상수에만 사용하는군요! 반영하여 수정하겠습니다.

f-lab-dev commented on src/main/java/com/flab/shoeauction/common/config/CacheProperties.java:
```diff
@@ -0,0 +1,28 @@
+ * 트래픽에 따라 캐시 별 만료기간을 쉽게 변경할 수 있도록 properties에서 cache 이름별 만료 기간을 Map에 바인딩한다.
+ */
+@Getter
+@Setter
```
이거는 꼭 setter를 넣어줘야하는걸까요? 생성자론 바인딩이 되지 않나요?

Hosick replied:
setter를 제거하고, 생성자를 이용해서 바인딩하는 방식으로 변경하였습니다!

Hosick committed:
feat: 리뷰 반영

-properties 바인딩을 수정자에서 생성자를 사용하는 방식으로 변경

issue number: #30

Hosick committed:
Merge branch 'develop' into feature/30

Hosick committed:
feat: 권한 체크 추가

- 상품, 브랜드 관리자 권한 체크 적용
- 통화, 사이즈 분류, 사이즈 단위 목록 조회 구현

issue number: #30
